### PR TITLE
Added body operations constructShapeFromBody() and constructMarkerFromBody().

### PR DESCRIPTION
**Pull request by peci1**
This helps very much when you want to visualize bodies in RViz and want to be sure the bug is in your code and not in the way you create the marker.

This can be backported to melodic. The ABI-breaking functions getScaledDimensions() could be at least added as non-virtual to keep ABI.

### CODE REVIEW

**Comment by v4hn** [comment]
@peci1 could you please rebase this to get conflicts resolved.

**Comment by peci1** [comment]
Rebased.

**Review by rhaschke — APPROVED** [review]

**Comment by rhaschke** [comment]
@v4hn, do you want to give this a second review?

**Review by v4hn — APPROVED** [review]
One more cleanup from my side.
If CI succeeds, I approve.

**Comment by peci1** [comment]
Should I prepare the backporting PR for melodic with non-virtual functions?

**Comment by v4hn** [comment]
If you are interested in the methods in melodic, then go ahead, yes.
I guess you propose that people can handle the down-cast from `bodies::Body` on their own before using the methods?

**Comment by peci1** [comment]
That's one way. Another way would be to also add the non-virtual methods to `bodies::Body` and switch using dynamic_casts or typeid. It'd be ugly, but probably not that much as there are only a few bodies defined...

**Comment by v4hn** [comment]
Sounds reasonable. At least it would provide the same API.
The (minor) downside is that the new interface would require RTTI.
But I don't know whether anyone is actually interested in disabling it nowadays.

**Comment by peci1** [comment]
Hmm, this could probably be handled by some CMake magic to detect RTTI and then making two different implementations (the non-RTTI one probably throwing an exception?). But I'm not sure if it's worth the hassle...

**Comment by v4hn** [comment]
> But I'm not sure if it's worth the hassle...

I agree and would just ignore it. Your proposal sounds good to include in melodic.

**Comment by rhaschke** [comment]
Is it really worth the effort to backport this? Which feature in the Melodic branches rely on it?

**Comment by peci1** [comment]
It'd at least allow me to drop my ugly custom workarounds in http://github.com/peci1/robot_body_filter/blob/7b69ec6d5a9231adc7eb0b52c9c270903522a21c/src/utils/bodies.cpp .

**Comment by rhaschke** [comment]
As long as you are the only one requiring this feature, you can simply build from master source, can't you?
UPDATE: However, if you _want_ to put this extra effort, you are, of course, welcome to do so.

**Comment by v4hn** [comment]
> As long as you are the only one requiring this feature, you can simply build from master source, can't you?

I don't see why you would try to convince contributors not to provide tested backports.
With MTC you are always the one who requires compatibility with the releases and downstream compatibility is always so much easier if upstream provides the same APIs across versions...

**Comment by rhaschke** [comment]
We always need to balance the extra effort for backporting (and here code different from the master branch is required to maintain ABI stability) and the benefit. I don't yet see other users than Martin utilizing the new feature in Melodic. I'm not trying to convince him to not provide the backport as I hopefully clarified in my update...

**Comment by peci1** [comment]
It's not just me who's using it. The package is released into melodic and has at least three users (or people who tried it and reported issues). Thus building against master is not an option.

The extra effort is worth it for me, so I'll prepare the PR - it will just take some time because of vacations...

**Comment by peci1** [comment]
And I think the logic is twisted:

> I don't yet see other users than Martin utilizing the new feature in Melodic

How could people use the new features when it has only been merged into Noetic? And there's more - I think the non-presence of `getScaledDimensions()` might have had discouraged some people from using geometric_shapes, because without this function you basically create bodies whose dimensions are unobservable from outside the library. And that makes utilizing these bodies pretty hard in downstream code (and that's the reason why I needed the ugly `#define private public` hack).

**Comment by rhaschke** [comment]
People can use this feature even if they are not using a (not yet existing) binary Noetic release: They just need to build this package from source within their ROS workspace.

**Comment by v4hn** [comment]
@peci1 As this just came up again in https://github.com/PR2/robot_self_filter/issues/23#issuecomment-674032618 ,
feel free to provide a melodic backport.

**Comment by peci1** [comment]
It took me time, but the backport is here: #209.

**Pull request merged**
